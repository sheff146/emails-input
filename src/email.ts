export class Email {
    private readonly _value: string;
    private readonly _isValid: boolean;

    get value(): string {
        return this._value;
    }

    get isValid(): boolean {
        return this._isValid;
    }

    constructor(value: string) {
        this._value = value;
        this._isValid = isValidEmail(value);
    }
}

function isValidEmail(value: string): boolean {
    return /./.test(value);
}
