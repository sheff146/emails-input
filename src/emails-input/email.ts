export class Email {
  private readonly _value: string;
  private readonly _isValid: boolean;

  get value(): string {
    return this._value;
  }

  get isValid(): boolean {
    return this._isValid;
  }

  constructor(value: string) {
    this._value = value;
    this._isValid = isValidEmail(value);
  }
}

// Taken from https://emailregex.com/
// Works 99.99%
const EMAIL_REGEX = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

function isValidEmail(value: string): boolean {
  return EMAIL_REGEX.test(value);
}
