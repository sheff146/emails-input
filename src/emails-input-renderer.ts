import { CallbackFn } from './interfaces';
import { Email } from './email';

export class EmailsInputRenderer {
    private _wrapper: HTMLDivElement;

    constructor(private container: HTMLElement, private onChanges: CallbackFn<string>) {
        this._wrapper = this.createWrapper();
        container.appendChild(this._wrapper);
    }

    render(emails: Email[]) {
        emails.forEach(email => {
            const item = this.createEmailItem(email);
            this._wrapper.appendChild(item);
        });
    }

    private createWrapper(): HTMLDivElement {
        const wrapper = document.createElement('div');

        wrapper.className = 'wrapper';
        wrapper.innerHTML = `<input type="text" class="wrapper__input" placeholder="add more people..."/>`;

        wrapper.addEventListener('click', (e) => {
            const target = e.target as HTMLElement;
            if (target.className === 'wrapper__item-remove') {
                const email = target.parentElement?.dataset?.email;
                if (email) {
                    this.onChanges({ addedItems: [], removedItems: [email] });
                }
            }
        });

        wrapper.addEventListener('keyup', (e) => {
            const target = e.target as HTMLElement;

            if (target.className === 'wrapper__input') {
                if (e.key === 'Enter' || e.key === ',') {
                    this.processSubmit(target as HTMLInputElement);
                }
            }
        });

        wrapper.addEventListener('blur', (e) => {
            const target = e.target as HTMLElement;

            if (target.className === 'wrapper__input') {
                this.processSubmit(target as HTMLInputElement);
            }
        });

        wrapper.addEventListener('paste', (e) => {
            debugger;
            const target = e.target as HTMLElement;

            if (target.className === 'wrapper__input') {
                this.processSubmit(target as HTMLInputElement);
            }
        });

        return wrapper;
    }

    private processSubmit(target: HTMLInputElement) {
        const values = target.value
            .split(',')
            .map(v => v.trim())
            .filter(v => v);

        if (values.length) {
            this.onChanges({ addedItems: values, removedItems: [] });
        }

        target.value = '';
    }

    private createEmailItem(email: Email): HTMLDivElement {
        const item = document.createElement('div');

        item.dataset.email = email.value;
        item.classList.add('wrapper__item');

        if (!email.isValid) {
            item.classList.add('wrapper__item_invalid');
        }

        item.innerHTML = `<span class="wrapper__item-title">${email.value}</span><button class="wrapper__item-remove">Remove</button>`;

        return item;
    }
}
