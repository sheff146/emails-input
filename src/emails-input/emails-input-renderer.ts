import { CallbackFn } from "./interfaces";
import { Email } from "./email";

const emailsInputStyles = require("./emails-input.css");

const styles = {
  container: emailsInputStyles["container"],
  item: emailsInputStyles["container__item"],
  emailItem: emailsInputStyles["container__item_email"],
  invalidItem: emailsInputStyles["container__item_invalid"],
  itemTitle: emailsInputStyles["container__email"],
  itemRemove: emailsInputStyles["container__remove"],
  inputItem: emailsInputStyles["container__item_input"],
  input: emailsInputStyles["container__input"]
};

const SUBMIT_KEYS = ["Enter", ","];

export class EmailsInputRenderer {
  private readonly _container: HTMLElement;

  constructor(private wrapper: HTMLElement, private onChanges: CallbackFn<string>) {
    this._container = this.createContainer();
    wrapper.appendChild(this._container);
  }

  render(emails: Email[]) {
    let existingItems = Array.prototype.slice.call(
      this._container.querySelectorAll(`.${styles.emailItem}`)
    ) as HTMLElement[];

    existingItems = existingItems.filter(item => {
      const hasEmail = emails.some(email => email.value === item.dataset.email);
      if (!hasEmail) {
        this._container.removeChild(item);
      }

      return hasEmail;
    });

    const itemsToAdd: Email[] = [];
    emails.forEach(email => {
      const hasEmail = existingItems.some(item => email.value === item.dataset.email);
      if (!hasEmail) {
        itemsToAdd.push(email);
      }
    });

    const input = this._container.querySelector(`.${styles.inputItem}`);
    itemsToAdd.forEach(email => {
      const item = createEmailItem(email);
      this._container.insertBefore(item, input);
    });
    this._container.scrollTop = this._container.scrollHeight - this._container.clientHeight;
  }

  private createContainer(): HTMLElement {
    const container = document.createElement("ul");

    container.className = styles.container;
    container.innerHTML =
      `<li class="${styles.item} ${styles.inputItem}">` +
      `<input type="text" class="${styles.input}" placeholder="add more people..."/>` +
      `</li>`;

    container.addEventListener("click", this.processButtonEvent.bind(this));
    container.addEventListener("keyup", this.processInputEvent.bind(this));
    container.addEventListener("blur", this.processInputEvent.bind(this), true);
    container.addEventListener("paste", this.processInputEvent.bind(this));

    return container;
  }

  private processButtonEvent(e: Event) {
    const target = e.target as HTMLElement;

    // istanbul ignore else
    if (target.classList.contains(styles.itemRemove)) {
      // istanbul ignore next
      const email = target.parentElement?.dataset?.email;
      // istanbul ignore else
      if (email) {
        this.onChanges({ addedItems: [], removedItems: [email] });
      }
    }
  }

  private processInputEvent(e: Event) {
    const target = e.target as HTMLInputElement;
    if (!target.classList.contains(styles.input)) {
      return;
    }

    switch (e.type) {
      case "keyup":
        if (SUBMIT_KEYS.indexOf((e as KeyboardEvent).key) >= 0) {
          this.processSubmit(target);
        }
        break;
      case "blur":
        this.processSubmit(target);
        break;
      case "paste":
        setTimeout(() => {
          this.processSubmit(target);
        }, 0);
    }
  }

  private processSubmit(target: HTMLInputElement) {
    // Split, trim, and filter empty values
    const values = target.value
      .split(",")
      .map(v => v.trim())
      .filter(v => v);

    if (values.length) {
      this.onChanges({ addedItems: values, removedItems: [] });
    }

    target.value = "";
  }
}

function createEmailItem(email: Email): HTMLElement {
  const item = document.createElement("li");

  item.dataset.email = email.value;
  item.classList.add(styles.item);
  item.classList.add(styles.emailItem);

  if (!email.isValid) {
    item.classList.add(styles.invalidItem);
  }

  const titleLabel = `${email.value}, ${email.isValid ? "valid" : "invalid"}`;
  const buttonLabel = `Remove ${email.value}`;
  const buttonContent = require("!!html-loader!./remove.svg");

  item.innerHTML =
    `<span class="${styles.itemTitle}" aria-label="${titleLabel}">${email.value}</span>` +
    `<button class="${styles.itemRemove}" aria-label="${buttonLabel}">${buttonContent}</button>`;

  return item;
}
