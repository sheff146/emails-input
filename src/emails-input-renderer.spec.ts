import { EmailsInputRenderer } from "./emails-input-renderer";
import { Email } from "./email";

describe("Emails Input Renderer", () => {
  let renderer: EmailsInputRenderer;
  let onChangesMock = jest.fn();
  let emails: Email[];

  beforeEach(() => {
    jest.clearAllMocks();
    emails = [new Email("abcdef"), new Email("abcdef@ghijk.lm"), new Email("fedcba@ghijk.lm")];
    document.body.innerHTML = '<div id="input-container"></div>';
    renderer = new EmailsInputRenderer(
      document.getElementById("input-container") as HTMLElement,
      onChangesMock
    );
  });

  describe("(snapshots)", () => {
    it("should create empty input", () => {
      expect(document.body).toMatchSnapshot();
    });

    it("should render email items", () => {
      renderer.render(emails);
      expect(document.body).toMatchSnapshot();
    });

    it("should rerender email items", () => {
      const emails1 = emails;
      const emails2 = [
        new Email("abcdef"),
        new Email("fedcba@ghijk.lm"),
        new Email("12345@gmail.com")
      ];

      renderer.render(emails1);
      renderer.render(emails2);

      expect(document.body).toMatchSnapshot();
    });
  });

  describe("(API)", () => {
    const initRemoveClick = (email: string) => {
      const button = document.querySelector(
        `[data-email="${email}"] .wrapper__item-remove`
      ) as HTMLElement;
      const event = new MouseEvent("click", { bubbles: true });
      button.dispatchEvent(event);
    };

    const initSubmitByKey = (text: string, submitKey: "Enter" | ",") => {
      const input = document.querySelector(`.wrapper__input`) as HTMLInputElement;
      input.value = text;
      const event = new KeyboardEvent("keyup", { bubbles: true, key: submitKey });
      input.dispatchEvent(event);
    };

    const initSubmitByEvent = (text: string, eventName: "blur" | "paste") => {
      const input = document.querySelector(`.wrapper__input`) as HTMLInputElement;
      input.value = text;

      const event = new Event(eventName, { bubbles: true });
      input.dispatchEvent(event);

      if (eventName === "paste") {
        jest.runAllTimers();
      }
    };

    beforeEach(() => {
      jest.useFakeTimers();
    });

    it("should notify on remove click", () => {
      renderer.render(emails);

      initRemoveClick("abcdef");

      expect(onChangesMock).toBeCalledWith({ addedItems: [], removedItems: ["abcdef"] });
    });

    it("should not notify submit on empty input", () => {
      renderer.render(emails);

      initSubmitByKey("", "Enter");

      expect(onChangesMock).not.toBeCalled();
    });

    it.each(["Enter", ","])("should notify submit on %s", submitKey => {
      renderer.render(emails);

      initSubmitByKey("123", submitKey as "Enter" | ",");

      expect(onChangesMock).toBeCalledWith({ addedItems: ["123"], removedItems: [] });
    });

    it("should notify submit on blur event", () => {
      renderer.render(emails);

      initSubmitByEvent("123", "blur");

      expect(onChangesMock).toBeCalledWith({ addedItems: ["123"], removedItems: [] });
    });

    it("should parse and filter complex input on paste", () => {
      renderer.render(emails);

      initSubmitByEvent("  123,, , 456   , abcdef@gmail.com,", "paste");

      expect(onChangesMock).toBeCalledWith({
        addedItems: ["123", "456", "abcdef@gmail.com"],
        removedItems: []
      });
    });
  });
});
