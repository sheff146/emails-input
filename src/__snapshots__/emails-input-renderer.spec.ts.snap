// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emails Input Renderer (snapshots) should create empty input 1`] = `
<body>
  <div
    id="input-container"
  >
    <ul
      class="wrapper"
    >
      <li
        class="wrapper__item wrapper__item_input"
      >
        <input
          class="wrapper__input"
          placeholder="add more people..."
          type="text"
        />
      </li>
    </ul>
  </div>
</body>
`;

exports[`Emails Input Renderer (snapshots) should render email items 1`] = `
<body>
  <div
    id="input-container"
  >
    <ul
      class="wrapper"
    >
      <li
        class="wrapper__item wrapper__item_email wrapper__item_invalid"
        data-email="abcdef"
      >
        <span
          class="wrapper__item-title"
        >
          abcdef
        </span>
        <button
          aria-label="Remove abcdef"
          class="wrapper__item-remove"
        >
          test-file-stub
        </button>
      </li>
      <li
        class="wrapper__item wrapper__item_email"
        data-email="abcdef@ghijk.lm"
      >
        <span
          class="wrapper__item-title"
        >
          abcdef@ghijk.lm
        </span>
        <button
          aria-label="Remove abcdef@ghijk.lm"
          class="wrapper__item-remove"
        >
          test-file-stub
        </button>
      </li>
      <li
        class="wrapper__item wrapper__item_email"
        data-email="fedcba@ghijk.lm"
      >
        <span
          class="wrapper__item-title"
        >
          fedcba@ghijk.lm
        </span>
        <button
          aria-label="Remove fedcba@ghijk.lm"
          class="wrapper__item-remove"
        >
          test-file-stub
        </button>
      </li>
      <li
        class="wrapper__item wrapper__item_input"
      >
        <input
          class="wrapper__input"
          placeholder="add more people..."
          type="text"
        />
      </li>
    </ul>
  </div>
</body>
`;

exports[`Emails Input Renderer (snapshots) should rerender email items 1`] = `
<body>
  <div
    id="input-container"
  >
    <ul
      class="wrapper"
    >
      <li
        class="wrapper__item wrapper__item_email wrapper__item_invalid"
        data-email="abcdef"
      >
        <span
          class="wrapper__item-title"
        >
          abcdef
        </span>
        <button
          aria-label="Remove abcdef"
          class="wrapper__item-remove"
        >
          test-file-stub
        </button>
      </li>
      <li
        class="wrapper__item wrapper__item_email"
        data-email="fedcba@ghijk.lm"
      >
        <span
          class="wrapper__item-title"
        >
          fedcba@ghijk.lm
        </span>
        <button
          aria-label="Remove fedcba@ghijk.lm"
          class="wrapper__item-remove"
        >
          test-file-stub
        </button>
      </li>
      <li
        class="wrapper__item wrapper__item_email"
        data-email="12345@gmail.com"
      >
        <span
          class="wrapper__item-title"
        >
          12345@gmail.com
        </span>
        <button
          aria-label="Remove 12345@gmail.com"
          class="wrapper__item-remove"
        >
          test-file-stub
        </button>
      </li>
      <li
        class="wrapper__item wrapper__item_input"
      >
        <input
          class="wrapper__input"
          placeholder="add more people..."
          type="text"
        />
      </li>
    </ul>
  </div>
</body>
`;
